solr/core/src/test/org/apache/solr/update/processor/UpsertConditionTest.java

```diff
@@ -0,0 +1,617 @@
+    newDoc.setField("field", "left-alone");
+
+    condition.copyOldDocFields(oldDoc, newDoc);
+
+    assertThat(newDoc.getFieldValue("field"), is("left-alone"));
+    assertThat(newDoc.getFieldValue("other_field"), is("old-value"));
+    assertThat(newDoc.getFieldValue("also-copied"), is("also-copied"));
+  }
+
+  @Test
+  public void givenExistingPermanentDelete_whenCheckingShouldInsertOrUpsert() {
+    List<UpsertCondition> conditions = givenMultipleConditions();
+
+    SolrInputDocument oldDoc = new SolrInputDocument();
+    SolrInputDocument newDoc = new SolrInputDocument();
+
+    oldDoc.setField("compliance_reason", "delete");
+
+    assertThat(UpsertCondition.shouldInsertOrUpsert(conditions, oldDoc, newDoc), is(false));
+  }
+
+  @Test
+  public void givenExistingSoftDelete_whenCheckingShouldInsertOrUpsert() {
+    List<UpsertCondition> conditions = givenMultipleConditions();
+
+    SolrInputDocument oldDoc = new SolrInputDocument();
+    SolrInputDocument newDoc = new SolrInputDocument();
+
+    oldDoc.setField("compliance_reason", "soft_delete");
+    oldDoc.setField("old_field", "not-kept-from-old");
+    newDoc.setField("new_field", "kept-from-new");
+
+    assertThat(UpsertCondition.shouldInsertOrUpsert(conditions, oldDoc, newDoc), is(true));
+    assertThat(newDoc.getFieldValue("new_field"), is("kept-from-new"));
+    assertThat(newDoc.getFieldValue("compliance_reason"), is("soft_delete"));
+    assertThat(newDoc.getFieldValue("old_field"), nullValue());
+  }
+
+  @Test
+  public void givenNewSoftDelete_whenCheckingShouldInsertOrUpsert() {
+    List<UpsertCondition> conditions = givenMultipleConditions();
+
+    SolrInputDocument oldDoc = new SolrInputDocument();
+    SolrInputDocument newDoc = new SolrInputDocument();
+
+    oldDoc.setField("old_field1", "kept-from-old1");
+    oldDoc.setField("old_field2", "kept-from-old2");
+    newDoc.setField("compliance_reason", "soft_delete");
+
+    assertThat(UpsertCondition.shouldInsertOrUpsert(conditions, oldDoc, newDoc), is(true));
+    assertThat(newDoc.getFieldValue("old_field1"), is("kept-from-old1"));
+    assertThat(newDoc.getFieldValue("old_field2"), is("kept-from-old2"));
+    assertThat(newDoc.getFieldValue("compliance_reason"), is("soft_delete"));
+  }
+
+  @Test
+  public void givenExistingMetrics_whenCheckingShouldInsertOrUpsert() {
+    List<UpsertCondition> conditions = givenMultipleConditions();
+
+    SolrInputDocument oldDoc = new SolrInputDocument();
+    SolrInputDocument newDoc = new SolrInputDocument();
+
+    oldDoc.setField("metric1", "kept-from-old1");
+    oldDoc.setField("metric2", "kept-from-old2");
+    oldDoc.setField("metric3", "not-kept-from-old");
+
+    assertThat(UpsertCondition.shouldInsertOrUpsert(conditions, oldDoc, newDoc), is(true));
+    assertThat(newDoc.getFieldValue("metric1"), is("kept-from-old1"));
+    assertThat(newDoc.getFieldValue("metric2"), is("kept-from-old2"));
+    assertThat(newDoc.getFieldValue("metric3"), nullValue());
```

Comment:
I was wondering if it's possible to clear/remove a field from the NEW doc, given the OLD doc and conditions? Upsert looks like it can only ever append or change fields.

Reply:
Yeah I'd assumed we'd only want to append fields for upsert.  I guess it wouldn't be impossible to make it always use what's in the old doc, but I assumed we'd want a way to be able to actual edit fields.  Otherwise we'd have to delete the doc first always.

Could maybe just be a difference syntax to denote when the OLD value should always replace the NEW value.

Reply:
yeah when we first chatted about this, I thought there would be 4 ways to decide how to handle a field: 1. firstnotnull(new, old) 2. firstnotnull(old, new) 3. always old, 4. always new

But at that point, we weren't really considering much beyond this compliance use-case and the Instagram use-case of improving but not downgrading docs.